File: array/combination_sum_II.py
class Solution(object):
    def combinationSum2(self, candidates, target):
        """
        :type candidates: List[int]
        :type target: int
        :rtype: List[List[int]]
        """
        ans=[]
        candidates.sort()
        
        self.combs(candidates,0,[],target,ans)
        
        return ans
        
    def combs(self, nums, index, path, target, ans):
        
        #if target<0:
         #   return
        if target==0:
            ans.append(path)
            return

        
        
        for i in range(index, len(nums)):
            if nums[i]>target:
                break
            if i>index and nums[i]==nums[i-1]:
                continue
            
            self.combs(nums, i+1, path+[nums[i]], target-nums[i], ans)

File: backtracking/combination_sum_II.cpp
class Solution {
public:
    void getcomb(vector<int> &nums, vector<vector<int> > &ans, vector<int> &path, int n, int tar)
    {
        if(tar==0)
        ans.push_back(path);

        for(int i=n;i<nums.size() && tar>=nums[i];i++)
        {
            if(i==n || nums[i-1]!=nums[i])
            {
                path.push_back(nums[i]);
                getcomb(nums, ans, path, i+1, tar-nums[i]);
                path.pop_back();
        
                
            }
        }
        
    }
    
    vector<vector<int>> combinationSum2(vector<int>& nums, int tar) {
        int n=nums.size();
        sort(nums.begin(), nums.end());
        vector<vector<int> > ans;
        vector<int> path;
        getcomb(nums, ans, path,0, tar);
        return ans;
    }
};